modules/math/shared/src/main/scala/gsp/math/RadialVelocity.scala

```diff
@@ -3,70 +3,85 @@
+  *
+  * The selection of units is based on references to velocities as low as 20 cm/s
+  * https://en.wikipedia.org/wiki/Radial_velocity
+  */
+final case class RadialVelocity private (rv: Quantity[BigDecimal, MetersPerSecond]) {
+
+  def toDoubleKilometersPerSecond: Double = rv.to[Double, KilometersPerSecond].value
+
+  /**
+    * Converts the radial velocity to a Redshift, approximate
+    * a return value of None should be understood as an infinity Redshift
+    */
+  def toRedshift: Option[Redshift] =
+    if (rv.value.abs < RadialVelocity.CValue) {
+      val i = (rv / RadialVelocity.C).value
+      val t = (1 + i) / (1 - i)
+      Some(Redshift(BigDecimal.decimal(scala.math.sqrt(t.toDouble) - 1).round(rv.value.mc)))
+    } else None
 
   override def toString =
-    f"RadialVelocity($toDoubleKilometersPerSecond%3.3fkm/s)"
-
+    s"RadialVelocity(${rv.to[Double, KilometersPerSecond].show})"
 }
 
 object RadialVelocity {
 
-  val fromMetersPerSecond: Prism[Int, RadialVelocity] =
-    Prism(Some(_: Int).filter(_.abs <= SpeedOfLight).map(new RadialVelocity(_) {}))(_.toMetersPerSecond)
-
-  val fromKilometersPerSecond: Format[BigDecimal, RadialVelocity] =
-    Format.fromPrism(fromMetersPerSecond)
-      .imapA(
-        n => new java.math.BigDecimal(n).movePointLeft(3),
-        d => d.underlying.movePointRight(3).intValue
-      )
-
-  /** Radial velocity of zero. */
-  val Zero: RadialVelocity = fromMetersPerSecond.unsafeGet(0)
-
-  /** Construct a [[RadialVelocity]] from floating point kilometers per second. */
-  def unsafeFromDoubleKilometersPerSecond(kms: Double): RadialVelocity =
-    fromKilometersPerSecond.unsafeGet(BigDecimal(kms))
-
-  def unsafeFromRedshift(z: Double): RadialVelocity = {
-    val C = SpeedOfLight.toDouble
-    val m = C * ((z + 1.0) * (z + 1.0) - 1.0) / ((z + 1.0) * (z + 1.0) + 1.0)
-    fromMetersPerSecond.unsafeGet(m.toInt)
-  }
-
-  /** Instances are ordered by their `.toMetersPerSecond` values. */
-  implicit val RadialVelocityOrder: Order[RadialVelocity] =
-    Order.by(_.toMetersPerSecond)
-
-  implicit val RadialVelocityShow: Show[RadialVelocity] =
+  // Reference: https://www.nist.gov/si-redefinition/meet-constants
+  // Exact value of the speed of light in m/s
+  private val CValue: BigDecimal = BigDecimal.decimal(SpeedOfLight.toDouble, MathContext.DECIMAL64)
+
+  val C: Quantity[BigDecimal, MetersPerSecond] =
+    CValue.withUnit[MetersPerSecond].toUnit[MetersPerSecond] // Speed of light in m/s
+
+  val fromMetersPerSecond: Prism[BigDecimal, RadialVelocity] =
+    Prism[BigDecimal, RadialVelocity](b =>
+      Some(b).filter(_.abs <= CValue).flatMap(v => RadialVelocity(v.withUnit[MetersPerSecond]))
+    )(_.rv.value)
+
+  /**
+    * Construct a RadialVelocity if the value is in the allowed range
+    * @group Constructors
+    */
+  def apply(rv: Quantity[BigDecimal, MetersPerSecond]): Option[RadialVelocity] =
+    if (rv.value.abs < CValue) Some(new RadialVelocity(rv)) else None
```

Comment:
I think we can drop the `new`, fwiw, since we no longer have `sealed abstract`.

Reply:
In this case we need `new` or it will try to call itself as `apply`